src/dune_pkg/opam_solver.ml

```diff
@@ -800,9 +800,7 @@ let reject_unreachable_packages =
             [ "name", Package_name.to_dyn name ]
         | Some (pkg : Lock_dir.Pkg.t), None -> Some (List.map pkg.depends ~f:snd)
         | None, Some (pkg : Local_package.For_solver.t) ->
-          let deps =
-            List.map pkg.dependencies ~f:(fun (d : Package_dependency.t) -> d.name)
-          in
+          let deps = Local_package.For_solver.dependency_names pkg in
```

Comment:
Isn't this wrong? You should not consider packages that haven't been selected for the reachability check. This would mean that you're including packages in the lock directory that are not needed for the actual build. This is a good thing to add a test for as well.

Reply:
Yes, you're completely right, thanks for pointing this out. This code needs to be changed because it uses the input to the solver as the output instead of using the solution that the solver outputs.

I'll add a test that checks that dependencies that don't get picked by the solver don't end up in the lock directory.

Reply:
I'm looking what the semantics of `reachable` is and according to #10814 it is about removing `post` dependencies (and also `post-deps-solving.t` is the only test failing if I comment out the reachability test).

Since this is basically about filtering the solver output via the possible deps from the input, how about still collecting all possible packages (since they are potentially reachable) except those that are `post` and using that for reachability?

That way packages that the solver is not picking will not be part of the lock dir (as the solution just does not contain them) and `post` packages won't be picked either (as they will be filtered out by the reachability check). This makes `post` a bit special but `post` is already a magic variable in `opam`.

Reply:
Added a test showing an unnecessary post dependencies being eliminated.

> how about still collecting all possible packages (since they are potentially reachable) except those that are post and using that for reachability?

I think it's just simpler to do this without any hacks. We've already resolved the `or` deps when writing the dependency list in the lock directory. Might as well just use that here instead of the formula

Reply:
In particular, I'm thinking of just using this snippet to evaluate the formula:

```
      Resolve_opam_formula.filtered_formula_to_package_names
        ~with_test:false
        (add_self_to_filter_env opam_package (Solver_env.to_env solver_env))
        version_by_package_name
        what
```

We already do this for external packages. No reason why we can't do it for local packages for the reachability check. You might want to wait until #11022 is merged though, as I'm going to move this snippet somewhere where it can be shared.

Reply:
Ah, I missed the comment. Ok, I'll try to evaluate the formula instead.

Reply:
Looking at the documentation of `filtered_formula_to_package_names` makes me question whether this is the right thing to use:

> Given a map associating package names with the version of that package in a solution, this function evaluates an `OpamFormula.t` to a concrete list of package names all contained in the solution with versions which satisfy the formula, or an error if the the version constraints or packages in the formula can't be satisfied with the packages in the solution. If the formula contains disjunctions (the `|` operator in opam) then there may be multiple sets of packages that satisfy the formula. In this case the choice of which set to return is made arbitrarily.

This sounds to me that if the solver chooses one side of the disjunction and `filtered_formula_to_package_names` the other, this will lead to all the solver-selected legitimate dependencies being filtered out. Or, if I pass in the result from the solver as the package name map, the function will fail to find a solution.